Suggests changes for database configuration topic (#10054)
keycloak/keycloak: docs/guides/src/main/server/db.adoc

By default, the server uses the `dev-file` database. This is the default database that the server will use to persist data and
only exists for development use-cases. The `dev-file` database is not suitable for production use-cases, and must be replaced before deploying to production.

> By default, the Keycloak server uses the `dev-file` database to persist data. This database is intended for development environments only. It is unsuitable for a production environment and must be replaced before deploying to production.

Reviewer: Should this substitute the paragraph before? If so, please delete the old paragraph, now it seems kind of doubled.

Reviewer: (the untouched white paragraph and the new green one i mean)
![Screenshot 2022-02-10 at 12 18 39](https://user-images.githubusercontent.com/89905860/153396798-5adb4074-1a53-4e88-941b-cb4d53ffaced.png)

By default, the server uses the `dev-file` database. This is the default database that the server will use to persist data and
only exists for development use-cases. The `dev-file` database is not suitable for production use-cases, and must be replaced before deploying to production.

By default, the Keycloak server uses the `dev-file` database to persist data. This database is intended for development environments only. It is unsuitable for a production environment and must be replaced before deploying to production.

== Configuring a database

For each supported database, the server provides some opinionated defaults to simplify database configuration. You complete the configuration by providing some key settings such as the database host and credentials.

> .Procedure

Reviewer 2: We are not using the `.Procedure` heading anywhere else in the guides, so please remove this

* Descriptions of objects

> Otherwise, characters are limited to those contained in database encoding, which is often 8-bit. However, for some database systems, you can enable UTF-8 encoding of Unicode characters and use the full Unicode character set in all text fields. For a given database, this choice might result in a shorter string length than the maximum string length supported by 8-bit encodings.

Reviewer 3: ```suggestion
Otherwise, characters are limited to those contained in database encoding, which is often 8-bit. However, for some database systems, you can enable UTF-8 encoding of Unicode characters and use the full Unicode character set in all text fields. For a given database, this choice might result in a shorter maximum string length than the maximum string length supported by 8-bit encodings.
```

=== Configuring Unicode support for an Oracle database

Unicode characters are supported in an Oracle database if the database was created with Unicode support in the VARCHAR and CHAR fields. For example, you configured AL32UTF8 as the database character set. In this case, the JDBC driver requires no special settings.

> If the database character set was not created with Unicode support, you need to configure the JDBC driver to support Unicode characters in the special fields. You configure two properties. Note that you can configure these properties as system properties or as connection properties.

Reviewer 3: It is the database, not character set, that needs to support Unicode.
```suggestion
If the database was not created with Unicode support, you need to configure the JDBC driver to support Unicode characters in the special fields. You configure two properties. Note that you can configure these properties as system properties or as connection properties.
```

. Set `oracle.jdbc.defaultNChar` to `true`.
+
[NOTE]
====
Setting this property might negatively impact performance. For more information, see the Oracle JDBC driver configuration documentation.
====

> . Optionally, set `oracle.jdbc.convertNcharLiterals` to `true`.

Reviewer 3: The new text does not explain the meaning of the properties nor reference to Oracle documentation for their explanation in the rewritten text. The Oracle doc for performance in the `[NOTE]` seems to point only to performance aspects of that property settings.

The correct rephrasing needs to refer to Oracle JDBC driver doc both for explanation of the meaning of the two properties and the performance impact.

=== Unicode support for a Microsoft SQL Server database

Unicode characters are supported only for the special fields for a Microsoft SQL Server database. The JDBC driver and database require no special settings.

> === Configuring Unicode support for a MySQL database

Reviewer 3: Is the difference between 
> **Configuring** Unicode support for a MySQL database

here and
> Unicode support for a Microsoft SQL Server database

a few lines above intentional (note the missing "Configuring")?

Author: @hmlnarik  I agree the headings are inconsistent, but the goal is to be consistent with topic.  "Configuring" Unicode support would not match the topic, Microsoff SQL Server database, where configuring does not apply.  However, am I missing your point somehow?

Reviewer 3: That makes sense, thank you. I only wanted to ensure that this was intentionally inconsistent.

> Unicode characters are supported in a MySQL database if the database was created with Unicode support in the VARCHAR and CHAR fields when using the CREATE DATABASE command. In other words, `utf8` was selected as the default database character set in MySQL 5.5.

Reviewer 3: This rewording is misleading. `utf8` is just one of the options for Unicode support in MSSQL, and while still valid, it is deprecated now. Also the version 5.5 is an example of MySQL version, not the one that is in any way suggested to be used, and now is rather old. That was the reason for the "e.g." in the original text.

> The utf8mb4 character set is not supported due to different storage requirements for the utf8 character set [1]). In that situation, the length restriction on non-special fields does not apply because columns are created to accommodate the number of characters, not bytes.  If the database default character set does not allow Unicode storage, only the special fields allow storing Unicode values.

Reviewer 3: What does the [1] mean? Is it https://dev.mysql.com/doc/refman/8.0/en/alter-table.html#alter-table-character-set or https://issues.jboss.org/browse/KEYCLOAK-3873 that was [in the original text](https://github.com/keycloak/keycloak-documentation/blob/567ca00fe12dbfd22915e85f0cb7f6d53d67cbd9/topics/database/unicode-considerations.adoc)?

```suggestion
The utf8mb4 character set is not supported due to different storage requirements for the utf8 character set, see MySQL documentation for details. In that situation, the length restriction on non-special fields does not apply because columns are created to accommodate the number of characters, not bytes.  If the database default character set does not allow Unicode storage, only the special fields allow storing Unicode values.
```

> . Start MySQL Server 5.5.

Reviewer 3: Why 5.5?

* Descriptions of objects

Otherwise, characters are limited to those contained in database encoding, which is often 8-bit. However, for some database systems, you can enable UTF-8 encoding of Unicode characters and use the full Unicode character set in all text fields. For a given database, this choice might result in a shorter string length than the maximum string length supported by 8-bit encodings.

> === Configuring Unicode support for an Oracle database

Reviewer 2: This section is just a copy/paste from the old documentation. @hmlnarik could you take a look at the suggested changes here?

Author: @hmlnarik Do you have any comment on this change?

Reviewer 3: I believe all my comments have been addressed by now. Thank you

Unicode characters are supported in an Oracle database if the database was created with Unicode support in the VARCHAR and CHAR fields. For example, you configured AL32UTF8 as the database character set. In this case, the JDBC driver requires no special settings.

If the database character set was not created with Unicode support, you need to configure the JDBC driver to support Unicode characters in the special fields. You configure two properties. Note that you can configure these properties as system properties or as connection properties.

> .Procedure

Reviewer 2: Same as above. Please remove as it is inconsistent. If we want to use .Procedure as a heading that should be proposed as a general thing, but personally I don't really see the case for this.

. Set `oracle.jdbc.defaultNChar` to `true`.
+
[NOTE]
====
Setting this property might negatively impact performance. For more information, see the Oracle JDBC driver configuration documentation.
====

. Optionally, set `oracle.jdbc.convertNcharLiterals` to `true`.

=== Unicode support for a Microsoft SQL Server database

Unicode characters are supported only for the special fields for a Microsoft SQL Server database. The JDBC driver and database require no special settings.

=== Configuring Unicode support for a MySQL database

Unicode characters are supported in a MySQL database if the database was created with Unicode support in the VARCHAR and CHAR fields when using the CREATE DATABASE command. In other words, `utf8` was selected as the default database character set in MySQL 5.5.

The utf8mb4 character set is not supported due to different storage requirements for the utf8 character set [1]). In that situation, the length restriction on non-special fields does not apply because columns are created to accommodate the number of characters, not bytes.  If the database default character set does not allow Unicode storage, only the special fields allow storing Unicode values.

> .Procedure

Reviewer 2: Same as above
